Signal/src/views/RemoteVideoView.m

```diff
@@ -0,0 +1,219 @@
+
+// As of RTC M61, iOS8 crashes when ending calls while de-alloc'ing the EAGLVideoView.
+// WebRTC doesn't seem to support iOS8 - e.g. their Podfile requires iOS9+, and they
+// unconditionally require MetalKit on a 64bit iOS8 device (which crashes).
+// Until WebRTC supports iOS8, we show a "upgrade iOS to see remote video" view
+// to our few remaining iOS8 users
+@interface NullVideoRenderer : UIView <RTCVideoRenderer>
+
+@end
+
+@implementation NullVideoRenderer
+
+- (instancetype)initWithFrame:(CGRect)frame
+{
+    self = [super initWithFrame:frame];
+    if (!self) {
+        return self;
+    }
+
+    self.backgroundColor = UIColor.blackColor;
+
+    UILabel *label = [UILabel new];
+    label.numberOfLines = 0;
+    label.text
+        = NSLocalizedString(@"CALL_REMOTE_VIDEO_DISABLED", @"Text shown on call screen in place of remote video");
+    label.textAlignment = NSTextAlignmentCenter;
+    label.font = [UIFont ows_boldFontWithSize:ScaleFromIPhone5(20)];
+    label.textColor = UIColor.whiteColor;
+
+    [self addSubview:label];
+    [label autoVCenterInSuperview];
+    [label autoPinWidthToSuperviewWithMargin:ScaleFromIPhone5(16)];
+
+    return self;
+}
+
+#pragma mark - RTCVideoRenderer
+
+/** The size of the frame. */
+- (void)setSize:(CGSize)size
+{
+    // Do nothing.
+}
+
+/** The frame to be displayed. */
+- (void)renderFrame:(nullable RTCVideoFrame *)frame
+{
+    // Do nothing.
+}
+
+@end
+
+@interface RemoteVideoView () <RTCEAGLVideoViewDelegate>
+
+@property (nonatomic, readonly) __kindof UIView<RTCVideoRenderer> *videoRenderer;
+
+// Used for legacy EAGLVideoView
+@property (nullable, nonatomic) NSMutableArray<NSLayoutConstraint *> *remoteVideoConstraints;
+
+@end
+
+@implementation RemoteVideoView
+
+- (instancetype)init
+{
+    self = [super init];
+    if (!self) {
+        return self;
+    }
+
+    // On iOS8: prints a message saying the feature is unavailable.
+    if (!SYSTEM_VERSION_GREATER_THAN_OR_EQUAL_TO(9, 0)) {
+        _videoRenderer = [NullVideoRenderer new];
+        [self addSubview:_videoRenderer];
+        [_videoRenderer autoPinEdgesToSuperviewEdges];
+    }
+
+// Currently RTC only supports metal on 64bit machines
+#if defined(RTC_SUPPORTS_METAL)
+    // On 64-bit, iOS9+: uses the MetalKit backed view for improved battery/rendering performance.
+    if (_videoRenderer == nil) {
+
+        // It is insufficient to check the RTC_SUPPORTS_METAL macro to determine Metal support.
+        // RTCMTLVideoView requires the MTKView class, available only in iOS9+
+        // So check that it exists before proceeding.
+        if ([MTKView class]) {
+            _videoRenderer = [[RTCMTLVideoView alloc] initWithFrame:CGRectZero];
+            [self addSubview:_videoRenderer];
+            [_videoRenderer autoPinEdgesToSuperviewEdges];
+            // HACK: Although RTCMTLVideo view is positioned to the top edge of the screen
+            // It's inner (private) MTKView is below the status bar.
```

Comment:
So the `MTKView` has constraints but they're wrong?  Seems like we should remove those constraints or make sure that the constraints we add have very high priority.

Reply:
Still curious about this.

Reply:
I made their priority explicitly "required".